File: 다이내믹프로그래밍/warshall.py
arr=[
[0, 1, 1, 1, 0, 0, 0],
[0, 0, 0, 1, 1, 0, 0],
[0, 0, 0, 0, 0, 1, 0],
[0, 0, 1, 0, 0, 1, 1],
[0, 0, 0, 1, 0, 0, 1],
[0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 1, 0]
]
def warshall(size):
    global arr
    for i in range(size):
        for j in range(size):
            for k in range(size):
                if i==j or i==k: continue
                if arr[i][j]==1 and arr[k][i]==1:
                    arr[k][j]=1
size=len(arr)
print("Adjacency Matrix ->")
for i in range(size):
    for j in range(size):
        print(arr[i][j],end=' ')
    print()
warshall(size)
print("Transitive Closure ->")
for i in range(size):
    for j in range(size):
        print(arr[i][j],end=' ')
    print()

File: 다이내믹프로그래밍/warshall.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
int arr[7][7] = {
	0, 1, 1, 1, 0, 0, 0,
	0, 0, 0, 1, 1, 0, 0,
	0, 0, 0, 0, 0, 1, 0,
	0, 0, 1, 0, 0, 1, 1,
	0, 0, 0, 1, 0, 0, 1,
	0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 1, 0};
void warshall(int);
int main(void){
    int size=sqrt(sizeof(arr)/sizeof(int));
    printf("Adjacency Matrix ->\n");
    for(int i=0;i<size;i++){
        for(int j=0;j<size;j++){
            printf("%d ",arr[i][j]);
        }
        printf("\n");
    }
    warshall(size);
    printf("Transitive Closure ->\n");
    for(int i=0;i<size;i++){
        for(int j=0;j<size;j++){
            printf("%d ",arr[i][j]);
        }
        printf("\n");
    }
    return 0;
}
void warshall(int size){
    for(int i=0;i<size;i++){
        for(int j=0;j<size;j++){
            for(int k=0;k<size;k++){
                if(i==j || i==k)continue;
                if(arr[i][j]==1&&arr[k][i]==1)arr[k][j]=1;
            }
        }
    }
}